rust/kernel/miscdev.rs

```diff
@@ -85,8 +85,11 @@ impl<T: Sync> FileOpenAdapter for Registration<T> {
     type Arg = T;
 
     unsafe fn convert(_inode: *mut bindings::inode, file: *mut bindings::file) -> *const Self::Arg {
+        // TODO: `SAFETY` comment required here even if `unsafe` is not present,
+        // because `container_of!` hides it. Ideally we would not allow
+        // `unsafe` code as parameters to macros.
         let reg = crate::container_of!((*file).private_data, Self, mdev);
```

Comment:
The macro could introduce a temp outside of the `unsafe` block that stores the value casted to a raw ptr. Ideally no `$foo:expr` happen inside an `unsafe` block for macros.

Reply:
+1, opening an issue.

Reply:
I guess this could also be a lint with some logic like:
  - If inside an `unsafe fn`...
  - And there is a macro call outside an `unsafe` block...
  - And `unsafe_op_in_unsafe_fn` is enabled in this context...
  - And the expansion of the macro introduces `unsafe`...
  - And if the node/AST/tree of a parameter appears as-is...
  - Then check that node/AST/tree is not inside `unsafe`.

Easier said than done, and I guess I am forgetting a few things, but... :^)

If you think it makes sense, I can open an issue too.

Reply:
Maybe. I believe there has been discussion previously regarding "unsafe" hygiene, but I can't remember what came of it. I think it was on internals.rust-lang.org.

Reply:
I guess it does not hurt to open it to keep track.

Reply:
:+1:

Reply:
https://github.com/rust-lang/rust-clippy/issues/7323